src/backend/distributed/commands/multi_copy.c

```diff
@@ -2265,6 +2273,15 @@ CitusCopyDestReceiverStartup(DestReceiver *dest, int operation,
 	copyDest->connectionStateHash = CreateConnectionStateHash(TopTransactionContext);
 
 	RecordRelationAccessIfReferenceTable(tableId, PLACEMENT_ACCESS_DML);
+
+	/*
+	 * For all the primary (e.g., writable) nodes, reserve a shared connection.
+	 * We do this upfront because we cannot know which nodes are going to be
+	 * accessed. Since the order of the reservation is important, we need to
+	 * do it right here. For the details on why the order important, see
+	 * the function.
+	 */
+	ReserveSharedConnectionCounterForAllPrimaryNodesIfNeeded();
```

Comment:
Could there be a scenario in which we already have reserved connections when we get here?

Reply:
Additional question: if we consider calling this function at the start of adaptive executor for anything other than single shard commands outside of a transaction blocks and always keep reservations until the end of the transaction block, how would that affect the design?

Reply:
> Could there be a scenario in which we already have reserved connections when we get here?

I think it could happen if a COPY can trigger another COPY such as [this](https://github.com/citusdata/citus/pull/4034/files#r460966908) one, yes we could have some reserved connections. And, in that case, we'd skip reserving the connection, which is wrong. 

So, we should better not have `HasAlreadyReservedConnection ()` check in `ReserveSharedConnectionCounterForAllPrimaryNodesIfNeeded ()` I guess?

> Additional question: if we consider calling this function at the start of adaptive executor for anything other than single shard commands outside of a transaction blocks and always keep reservations until the end of the transaction block, how would that affect the design?

Should be fine in terms of correctness, even running the regression tests with it pass. (We should reserve the connections after executor locks are acquired) 

Though, we should probably only do it for the worker nodes that are going to be accessed (e.g., `ReserveSharedConnectionCounterForNodeListIfNeeded (workerNodeList)`). Otherwise we could access to the shared memory unnecessarily. 

Overall, I'd expect some overhead of such an approach, we'd unnecessarily go through reservation process. We should  run some benchmarks if we  decide to do that. 

I find it not to have it for the executor simpler, though I can consider doing it the other way if you think so.